libs/dmlf/include/dmlf/type_map.hpp

```diff
@@ -0,0 +1,74 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include <atomic>
+#include <exception>
+#include <iostream>
+#include <unordered_map>
+
+namespace fetch {
+namespace dmlf {
+
+template <typename ValueType = std::string>
+class TypeMap
+{
+public:
+  using IntToValueMap = std::unordered_map<int, ValueType>;
+
+  template <typename KeyType>
+  ValueType find() const
+  {
+    auto iter = map.find(getTypeId<KeyType>());
+    if (iter == map.end())
+    {
+      throw std::runtime_error{"Type not registered"};
+    }
+    return iter->second;
+  }
+
+  template <typename KeyType>
+  // void put(ValueType &&value)
+  void put(ValueType value)
+  {
+    auto iter = map.find(getTypeId<KeyType>());
+    if (iter != map.end())
+    {
+      throw std::runtime_error{"Type already registered with name '" + iter->second + "'"};
+    }
+    // map[getTypeId<KeyType>()] = std::forward<ValueType>(value);
+    map[getTypeId<KeyType>()] = value;
+  }
+
+  template <typename KeyType>
+  inline static int getTypeId()
```

Comment:
This looks like a round about way to basically run `typid` to get a `std::type_info`. Maybe using that would provide a cleaner interface?

Reply:
@ejfitzgerald https://en.cppreference.com/w/cpp/language/typeid 
that is true.
using `std::type_index`